subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskContainerTest.groovy

```diff
@@ -1389,43 +1405,34 @@ class DefaultTaskContainerTest extends Specification {
         1 * taskFactory.create(_ as TaskIdentity) >> task("task")
     }
 
-    void "can remove lazy created task without breaking TaskProvider"() {
+    void "can remove eager created task and named provider update his state"() {
         given:
-        def customTask = task("task", CustomTask)
-        1 * taskFactory.create(_ as TaskIdentity, _ as Object[]) >> customTask
+        taskFactory.create(_ as TaskIdentity) >> task("task", CustomTask)
 
         and:
-        def createProvider = container.register("task", CustomTask)
+        def customTask = container.create("task", CustomTask)
 
         when:
-        container.remove(customTask)
+        def provider = container.named("task")
 
         then:
-        createProvider.get() == customTask
-        createProvider.present
+        provider.present
+        provider.get() == customTask
 
         when:
-        def getProvider = container.named("task")
+        def didRemoved = container.remove(customTask)
 
         then:
-        getProvider.get() == customTask
-        getProvider.present
-    }
-
-    void "can remove eager created task without breaking TaskProvider"() {
-        given:
-        taskFactory.create(_ as TaskIdentity) >> task("task", CustomTask)
-
-        and:
-        def customTask = container.create("task", CustomTask)
-        def getProvider = container.named("task")
+        didRemoved
+        !provider.present
+        provider.orNull == null
 
         when:
-        container.remove(customTask)
+        provider.get()
 
         then:
-        getProvider.get() == customTask
-        getProvider.present
+        def ex = thrown(IllegalStateException)
+        ex.message == "No value has been specified for this provider."
```

Comment:
Seems like we should provide a better error message here rather than delegating to the `AbstractProvider` message.  Something like "this provider has been removed" or "this provider is no longer present in its container".